lib/client.ml

```diff
@@ -51,22 +65,62 @@ let recv_next client =
   Lwt_mutex.unlock !client.recv_mutex;
   Lwt.return (msg_buffer, Peer.from_sockaddr addr)
 
-let serve client msg_handler =
+let next_request client = Tqueue.take_opt !client.request_inbox
+let next_response client = Tqueue.take_opt !client.response_inbox
+
+let request client request peer =
+  let%lwt () = send_to client request peer in
+  Tqueue.wait_to_take !client.response_inbox
+
+let route client peer msg router =
+  let open Message in
+  let msg = router peer msg in
+  match msg.label with
+  | Message.Request -> Tqueue.add msg !client.request_inbox
+  | Message.Response -> Tqueue.add msg !client.response_inbox
+
+let serve client router msg_handler =
   let rec server () =
-    let%lwt request, peer = recv_next client in
-    let%lwt () = Lwt_mutex.lock !client.state_mutex in
-    let response = msg_handler !client.state peer request in
-    let%lwt () = send_to client response peer in
-    Lwt_mutex.unlock !client.state_mutex;
+    let%lwt message, peer = recv_next client in
+    let%lwt () = route client peer message router in
+
+    let%lwt request = next_request client in
+
+    let%lwt () =
+      match request with
+      | Some request ->
+        let%lwt () = Lwt_mutex.lock !client.state_mutex in
```

Comment:
What is the behavior of the lock? Will it lock only the current thread?

Reply:
The lock stops any other thread from reading from/writing to the client's state until the lock is released

Reply:
Is it like an `mvar` which is under the hood of the [Concurrent.Lock](https://hackage.haskell.org/package/concurrent-extra-0.7.0.12/docs/Control-Concurrent-Lock.html) I already used in Haskell? It would help me if it is the case 😇